File: core/shading/shader_applicator.cpp
/*
    Beatmup image and signal processing library
    Copyright (C) 2019, lnstadrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "shader_applicator.h"
#include "../gpu/pipeline.h"
#include "../debug.h"

using namespace Beatmup;

bool ShaderApplicator::processOnGPU(GraphicPipeline &gpu, TaskThread &thread) {
    // distribute texture units over samplers
    {
        int unit = mainInput ? 1 : 0;
        for (auto input : samplers)
            if (input.first != ImageShader::INPUT_IMAGE_ID) {
                RuntimeError::check(input.second->getTextureFormat() != GL::TextureHandler::TextureFormat::OES_Ext,
                    "OES_EXT samplers are only supported when bound to '" + ImageShader::INPUT_IMAGE_ID + "' sampler variable of type " + ImageShader::INPUT_IMAGE_DECL_TYPE);
                shader->setInteger(input.first, unit);
                unit++;
            }
    }

    // prepare shader
    shader->prepare(gpu, mainInput, output);

    // bind textures
    {
        size_t unit = mainInput ? 1 : 0;
        for (auto input : samplers)
            if (input.first != ImageShader::INPUT_IMAGE_ID) {
                gpu.bind(*input.second, unit, TextureParam::INTERP_LINEAR);
                unit++;
            }
    }

    // process
    shader->process(gpu);
    return true;
}


void ShaderApplicator::beforeProcessing(ThreadIndex threadCount, ProcessingTarget target, GraphicPipeline *gpu) {
    NullTaskInput::check(output, "output bitmap");
    NullTaskInput::check(shader, "image shader");
    if (mainInput)
        readLock(gpu, mainInput, ProcessingTarget::GPU);
    for (auto _ : samplers)
        readLock(gpu, _.second, ProcessingTarget::GPU);
    writeLock(gpu, output, ProcessingTarget::GPU);
}


void ShaderApplicator::afterProcessing(ThreadIndex threadCount, GraphicPipeline* gpu, bool aborted) {
    if (mainInput)
        unlock(mainInput);
    for (auto _ : samplers)
        unlock(_.second);
    unlock(output);
}


ShaderApplicator::ShaderApplicator():
    shader(nullptr), mainInput(nullptr), output(nullptr)
{}


void ShaderApplicator::addSampler(AbstractBitmap* bitmap, const std::string uniformName) {
    if (uniformName == ImageShader::INPUT_IMAGE_ID)
        mainInput = bitmap;
    else
        samplers[uniformName] = bitmap;
}


bool ShaderApplicator::removeSampler(const std::string uniformName) {
    if (uniformName == ImageShader::INPUT_IMAGE_ID) {
        if (mainInput) {
            mainInput = nullptr;
            return true;
        }
        else
            return false;
    }

    auto sampler = samplers.find(uniformName);
    if (sampler == samplers.end()) {
        return false;
    }
    else {
        samplers.erase(sampler);
        return true;
    }
}


void ShaderApplicator::clearSamplers() {
    mainInput = nullptr;
    samplers.clear();
}


void ShaderApplicator::setOutputBitmap(AbstractBitmap *bitmap) {
    output = bitmap;
}


void ShaderApplicator::setShader(ImageShader *shader) {
    this->shader = shader;
}

File: android/lib/src/main/java/Beatmup/Shading/ShaderApplicator.java
/*
    Beatmup image and signal processing library
    Copyright (C) 2019, lnstadrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package Beatmup.Shading;

import android.util.ArrayMap;

import java.util.Map;

import Beatmup.Bitmap;
import Beatmup.Context;
import Beatmup.Task;

/**
 * A Task applying an image shader to bitmaps
 */
public class ShaderApplicator extends Task {
    private Map<String, Bitmap> samplers;
    private Bitmap output;
    private ImageShader shader;

    private static native long newShaderApplicator();
    private native void addSampler(long handle, Bitmap bitmap, String uniformName);
    private native void setOutput(long handle, Bitmap bitmap);
    private native void setShader(long handle, ImageShader shader);

    /**
     * Creates a new shader applicator.
     * @param context   A Beatmup context instance
     */
    public ShaderApplicator(Context context) {
        super(context, newShaderApplicator());
        samplers = new ArrayMap<>();
    }

    /**
     * Attaches an ImageShader to the applicator.
     * @param shader    The shader
     */
    public void setShader(ImageShader shader) {
        this.shader = shader;
        setShader(handle, shader);
    }

    /**
     * @return currently used ImageShader or null of not set
     */
    public ImageShader getShader() {
        return shader;
    }

    /**
     * Adds an image to sample.
     * @param input     The image
     * @param name      GLSL sample variable name the image is bound to
     */
    public void addSampler(Bitmap input, String name) {
        this.samplers.put(name, input);
        addSampler(handle, input, name);
    }

    /**
     * Adds the main image.
     * The image is bound to {@link ImageShader#INPUT_IMAGE_ID) variable} which supports {@link Beatmup.Android.ExternalBitmap} textures to be used directly.
     * @param input     The image
     */
    public void addSampler(Bitmap input) {
        addSampler(input, ImageShader.INPUT_IMAGE_ID);
    }

    /**
     * Sets the output image.
     * It stores the shader output result after the ShaderApplicator task is run.
     * @param output    The image
     */
    public void setOutput(Bitmap output) {
        this.output = output;
        setOutput(handle, output);
    }

    /**
     * @return the output image or null if not set.
     */
    public Bitmap getOutput() {
        return output;
    }
}
